File: services/anonymizer-service/internal/clients/ai_coordinator_client.go
package clients

import (
	"bytes"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"time"
)

// Define task types known by the coordinator
const (
	TaskTypeAnonymizeText = "anonymize_text"
	// Add other task types later: "moderate_text", "moderate_image", etc.
)

// AICoordinatorRequest defines the structure for requests TO the coordinator
type AICoordinatorRequest struct {
	TaskType string            `json:"task_type"`        // e.g., TaskTypeAnonymizeText
	Payload  interface{}       `json:"payload"`          // The data for the task (e.g., map[string]string{"text": "..."})
	Config   map[string]string `json:"config,omitempty"` // Optional: specific model hints, etc.
}

// AICoordinatorResponse defines the generic structure for responses FROM the coordinator
// The actual 'Result' will vary based on the task type.
type AICoordinatorResponse struct {
	Success bool        `json:"success"`
	Result  interface{} `json:"result,omitempty"` // Use map[string]interface{} or specific structs
	Error   string      `json:"error,omitempty"`
}

// AnonymizeTextResult defines the expected structure within the 'Result' field for anonymization tasks
type AnonymizeTextResult struct {
	AnonymizedText string `json:"anonymized_text"`
	// Add other fields returned by the specific AI adapter via the coordinator if needed
}

// AICoordinatorClient holds configuration
type AICoordinatorClient struct {
	BaseURL    string
	HttpClient *http.Client
}

// NewAICoordinatorClient creates a new client instance
func NewAICoordinatorClient(baseURL string) *AICoordinatorClient {
	return &AICoordinatorClient{
		BaseURL: baseURL,
		HttpClient: &http.Client{
			Timeout: 20 * time.Second, // AI tasks might take longer
		},
	}
}

// RequestAnonymization sends an anonymization task request to the AI Coordinator
func (c *AICoordinatorClient) RequestAnonymization(text string) (*AnonymizeTextResult, error) {
	// The payload specific to the anonymize_text task
	payload := map[string]string{"text": text}

	coordReq := AICoordinatorRequest{
		TaskType: TaskTypeAnonymizeText,
		Payload:  payload,
		// Config: map[string]string{"model_preference": "ollama-fast"}, // Example optional config
	}

	payloadBytes, err := json.Marshal(coordReq)
	if err != nil {
		log.Printf("Error marshalling AI coordinator request payload: %v", err)
		return nil, fmt.Errorf("failed to create AI coordinator request payload: %w", err)
	}

	// Assuming the coordinator has a single endpoint like /process
	reqUrl := fmt.Sprintf("%s/process", c.BaseURL)
	req, err := http.NewRequest(http.MethodPost, reqUrl, bytes.NewBuffer(payloadBytes))
	if err != nil {
		log.Printf("Error creating request to AI coordinator: %v", err)
		return nil, fmt.Errorf("failed to create AI coordinator request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := c.HttpClient.Do(req)
	if err != nil {
		log.Printf("Error sending request to AI coordinator at %s: %v", reqUrl, err)
		return nil, fmt.Errorf("AI coordinator request failed: %w", err)
	}
	defer resp.Body.Close()

	// Decode the generic response first
	var coordResp AICoordinatorResponse
	if err := json.NewDecoder(resp.Body).Decode(&coordResp); err != nil {
		log.Printf("Error decoding AI coordinator generic response (status %d): %v", resp.StatusCode, err)
		// Return error even if status is 200 but body is unparsable
		return nil, fmt.Errorf("failed to decode AI coordinator response: %w", err)
	}

	// Check for non-OK status codes OR explicit failure in the response body
	if resp.StatusCode != http.StatusOK || !coordResp.Success {
		errorMsg := coordResp.Error
		if errorMsg == "" {
			errorMsg = fmt.Sprintf("AI Coordinator returned status %d with success=false", resp.StatusCode)
		}
		log.Printf("AI Coordinator request failed: %s", errorMsg)
		return nil, fmt.Errorf("AI coordinator task failed: %s", errorMsg)
	}

	// If successful, parse the specific result type (AnonymizeTextResult)
	// The Result field is likely a map[string]interface{} after generic decoding,
	// so we marshal it back to bytes and unmarshal into the specific struct.
	resultBytes, err := json.Marshal(coordResp.Result)
	if err != nil {
		log.Printf("Error marshalling coordinator result field: %v", err)
		return nil, fmt.Errorf("failed to process coordinator result structure: %w", err)
	}

	var anonymizeResult AnonymizeTextResult
	if err := json.Unmarshal(resultBytes, &anonymizeResult); err != nil {
		log.Printf("Error unmarshalling specific AnonymizeTextResult from coordinator response: %v", err)
		return nil, fmt.Errorf("failed to decode specific anonymization result: %w", err)
	}

	log.Printf("Successfully received anonymization result via AI Coordinator.")
	return &anonymizeResult, nil
}

File: services/moderation-service/lib/aiCoordinatorClient.js
const AI_COORDINATOR_URL = process.env.AI_COORDINATOR_URL;
const REQUEST_TIMEOUT_MS = 20000; // 20 seconds timeout for AI tasks

if (!AI_COORDINATOR_URL) {
    console.error('FATAL ERROR: AI_COORDINATOR_URL environment variable is not set.');
    process.exit(1);
}

// Define task types (should match definitions in Go client/coordinator)
const TASK_TYPES = {
    MODERATE_TEXT: 'moderate_text',
    MODERATE_IMAGE: 'moderate_image',
    // Add other types as needed
};

/**
 * Sends a task request to the AI Coordinator service.
 * @param {string} taskType - The type of task (e.g., TASK_TYPES.MODERATE_TEXT).
 * @param {object} payload - The specific data for the task.
 * @param {object} [config] - Optional configuration for the task.
 * @returns {Promise<object>} - The result part of the AI Coordinator's response.
 * @throws {Error} - If the request fails or the coordinator returns an error.
 */
async function requestAiTask(taskType, payload, config = {}) {
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), REQUEST_TIMEOUT_MS);

    const requestBody = {
        task_type: taskType,
        payload: payload,
        ...(Object.keys(config).length > 0 && { config }), // Add config only if not empty
    };

    const url = `${AI_COORDINATOR_URL}/process`; // Assuming /process endpoint
    console.log(`Sending task '${taskType}' to AI Coordinator at ${url}`);

    try {
        const response = await fetch(url, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'Accept': 'application/json',
            },
            body: JSON.stringify(requestBody),
            signal: controller.signal, // Add abort signal for timeout
        });

        clearTimeout(timeoutId); // Clear timeout if fetch completes

        const responseBody = await response.json(); // Attempt to parse JSON regardless of status

        if (!response.ok || !responseBody.success) {
            const errorMessage = responseBody?.error || `AI Coordinator returned status ${response.status}`;
            console.error(`AI Coordinator task failed: ${errorMessage}`, responseBody);
            throw new Error(`AI Coordinator task '${taskType}' failed: ${errorMessage}`);
        }

        console.log(`Successfully received result for task '${taskType}' from AI Coordinator.`);
        return responseBody.result; // Return only the result part on success

    } catch (error) {
        clearTimeout(timeoutId); // Ensure timeout is cleared on error
        if (error.name === 'AbortError') {
            console.error(`AI Coordinator request timed out after ${REQUEST_TIMEOUT_MS}ms`);
            throw new Error(`AI Coordinator request timed out`);
        }
        console.error(`Error communicating with AI Coordinator: ${error.message}`);
        // Rethrow a generic error or the specific error
        throw new Error(`Failed to communicate with AI Coordinator: ${error.message}`);
    }
}

module.exports = {
    requestAiTask,
    TASK_TYPES,
};